rlpr-2.06/man/rlpq.1
.\" SCCS "@(#)rlpq.1	1.3	01/01/02 meem"
.TH rlpq 1 "1999/10/28" "rlpr 2.04" "UNIX Reference Manual"
.SH NAME
rlpq \- remote line printer queue examination program
.SH SYNOPSIS
\fBrlpq\fP [-lmqNV] [\fB-H\fPprinthost] [\fB-P\fPprinter] [\fB-X\fPproxyhost]
[\fB--debug\fP] [\fB--timeout=\fPseconds] [\fB--port=\fPport] [ \fBjob # ... \fP] [ \fBusername ... \fP]
.SH DESCRIPTION
.PP
\fBrlpq\fP uses TCP/IP to examine print queues for printers anywhere
on a network.  Unlike lpq, it does not require that the remote
printers be explicitly known to the local system (traditionally
through \fI/etc/printcap\fR), and thus is considerably more flexible
and requires less administration.
.PP
\fBrlpq\fP can be used anywhere a traditional \fBlpq\fP might be used,
and is backwards compatible with traditional BSD \fBlpq\fP.  If
\fBrlpq\fP is invoked as \fBlpq\fP, it preserves all known lpq
semantics.
.PP
\fBrlpq\fP can be set up, installed, and used in the same two ways
as \fBrlpr\fP(1) (that is, either with or without the \fBrlprd\fP(8)
proxy daemon).  Please see \fBrlpr\fP(1) for details, and to
see whether you will need to use the \fBrlprd\fP(8) proxy daemon or
not.
.PP
.SH OPTIONS
.PP
.SS Environment/Setup
.TP
.I \-\-debug
Print gobs of debugging information.
.TP
.I \-N, \-\-no-bind
Don't try to bind to privileged port before connecting to lpd.  Please
see the \fBCONFIGURING A PROXY\fP section in \fBrlpr\fP(1) for
information on use of this option.
.TP
.I \-\-port=number
Select an alternate port (instead of 7290) to connect to, if using
rlprd.  Usually not needed.
.TP
.I \-H, \-\-printhost=host
Select the host to print to (used with \fI-P\fR).

Instead of using \fI-H\fR, one can specify the hostname directly
including it with the printer name with the \fIprinter@hostname\fR
syntax.
.TP
.I \-P, \-\-printer=printername, \-\-queue=printername
Select the printer to print to (used with \fI-H\fR).
.TP
.I \-X, \-\-proxy=proxyhost, \-\-proxyhost=proxyhost
Select the proxy host to use, if necessary.
.TP
.I \-\-timeout=seconds
Set the inactivity timer.  If the connection hangs for more than
\fBseconds\fP seconds, then \fBrlpq\fP will give up.  Use the special
value `-1' to wait forever.  Default timeout is 3 seconds.
.TP
.I \-V, \-\-version
Print version and exit.
.SS Behavior
.TP
.I \-l, \-\-long
Produce a more detailed report.
.TP
.I \-m, \-\-mine
List only the jobs belonging to you.
.TP
.I \-q, \-\-quiet, \-\-silent
Quiet mode \- stay quiet (except for fatal errors).  When this flag is
set, rlpq's return value indicates whether there are any items in
the specified queue that meet the specified criteria (a return value
of zero indicates success).
.PP
.SH "ENVIRONMENT VARIABLES"
The following environment variables are used by \fBrlpq\fP:
.TP
.B RLPR_PRINTHOST
Specifies the default host's queue to examine.
.TP
\fBPRINTER\fP or \fBLPDEST\fP
Specifies the default printer to check on the host.
.br
First \fBPRINTER\fP is consulted, then \fBLPDEST\fP.

Note that one can also specify the host to examine by setting
printer to be \fIprinter@hostname\fR.  This may be more convenient
than setting \fBRLPR_PRINTHOST\fP, but will confuse the traditional
BSD print commands.
.TP
\fBRLPR_PROXYHOST\fP
Specifies a proxy host to use, if necessary.
.SH FILES
.nf
.ta \w'123456789012345678'u
~/.rlprrc	Personal printer/host database
/etc/hosts.lpd	Host-based security on printhost
/etc/hosts.equiv	Host-based security on printhost
/etc/passwd	Personal identification
/etc/rlprrc	System-wide printer/host database
.SH "SEE ALSO"
\fBrlpr\fP(1), \fBrlprm\fP(1), \fBrlprd\fP(8), \fBrlprrc\fP(5)
.SH AUTHOR
meem <meem@gnu.org>
.SH "BUGS / LIMITATIONS"
Due to the dynamic nature of print queues, it is possible that the
actual state of the queue may have changed between the time the
snapshot was generated and the output was produced.



rlpr-2.06/NEWS
Changes in release 2.06:

* fixed several autoconf-related problems introduced in 2.05.
* support fallback to syslog() when vsyslog() is not available.

Changes in release 2.05:

* fixed security flaw detailed at: http://www.osvdb.org/7194

Changes in release 2.04:

* removed -W flag so that it does not conflict with POSIX.2 systems.
* fixed rlpr to not expect an argument with --windows (nee -W) option.
* fixed `bsdlinks' target to not accidentally remove lprm.
* stubbed out i18n/l10n functions to make ./configure --disable-nls work.
* enhanced rlpr to print a warning if the user attempts to print an
  empty file.
  
Changes in release 2.03:

* added support for printer alternation in check-server.sh.
* fixed message displayed when incorrect non-long style options are
  passed to rlpr, rlpq, rlprd, or rlprm.
* added configurable configuration directory (RLPR_CONFDIR) support.
* added support for running `configure' out of other than `top_srcdir'
* fixed build problem that caused libnsl and libsocket (when required)
  not to appear on link line.
* new autoconf macros for EXIT_FAILURE/EXIT_SUCCESS.
* fixed two bugs in rlprrc lookups of hosts for a printer.
* fixed bug which prevented rlprrc lookups for a printer for a host
  (now, default to `lp' is only done when the rlprrc lookup fails).

Changes in release 2.02:

* fixed connect() to report the correct error message upon failure.
* changed Makefiles to never use += operator (for portability).
* revised code to compile correctly on newer releases of GNU libc.
* updated pointers to packages in README.
* fixed --no-bind to work when running as root.
* fixed Makefiles so error.o is now only put in librlpr once.
* fixed broken configure.in custom checks for MAXHOSTNAMELEN and off_t.
* added custom check for h_errno definition in <netdb.h>.
* fixed test(1) brokenness in check-server.sh.

Changes in release 2.01:

* fixed bug that would cause rlpr to send from unprivileged ports
* fixed some longstanding argument parsing problems
* added `--hostname' option
* fixed bug that caused rlpr to be interrupted with SIGPOLL
* added generic support for vendor extensions to lpr
* added support for some hp/ux lpr extensions
* changed the default timeout to be more reasonable
* small fixes to make the code compile properly on aix 4.1.x
* small fixes to make the code compile properly on freebsd
* made ~/.rlprrc truly optional

Changes in release 2.00:

* complete rewrite of rlpr, rlpq, and rlprd - much more modular
* addition of new rlprm command
* complete re-architecture of build environment (deep directories now)
* new hooks for internationalization support
* build procedure now uses automake in addition to autoconf
* remote printing commands now support a --timeout option in case
  remote end dies
* new --send-data-first option to rlpr to work with buggy lpd's
* rlpr now works with hp jetdirect printers
* rlpr now includes a shiny-new progress meter, useful when printing
  files over slow links
* all printing commands default to `lp' printer now, like bsd
* RLPR_PROXY has been removed; use RLPR_PROXYHOST instead
* uses automake, autoconf, etc..


rlpr-2.06/src/rlpq.c
/*
 * Copyright (c) 1998-1999 peter memishian (meem), meem@gnu.org
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 */

#pragma ident "@(#)rlpq.c	1.1	99/09/16 meem"

#include <sys/types.h>
#include <errno.h>
#include <pwd.h>
#include <stdlib.h>
#include <string.h>
#include <getopt.h>
#include <unistd.h>
#include <stdio.h>
#include <lib.h>

#include "intl.h"
#include "component.h"
#include "msg.h"
#include "util.h"
#include "client.h"
#include "rlprrc.h"
#include "rlpq.h"
#include "rfc1179.h"

static struct component *components[] =
{
    &comp_msg,
    &comp_component,
    &comp_client,
    &comp_rlpq,
    0
};

const char		*program_name;
const char		*bsd_program_name = "lpq";

static struct rlpr_rlpq	*rlpr_rlpq;

static const char	*make_queue_request(const char *, char * const *);

int
main(int argc, char *argv[])
{
    const char	       *req;
    struct component   *comp;
    char	       *user_argv[] = { 0, 0 };
    char * const       *user_list = user_argv;
    int			sock_fd;
    int			retval = EXIT_SUCCESS;

    program_name = argv[0];
    toggle_root();

    setlocale(LC_ALL, "");
    bindtextdomain(PACKAGE, LOCALEDIR);
    textdomain(PACKAGE);

    if ((comp = component_init(components, argc, argv)) != 0)
	msg(R_FATAL, 0, "component `%s': init() failed!", comp->name);

    /*
     * if --mine is specified, then use a two-element argv[] that
     * has the first element set to the current user
     */

    if (rlpr_rlpq->mine_only)
	user_argv[0] = rlpr_rlpq->user;
    else
	user_list    = &argv[optind];

    /*
     * make the queue request message (which we need to make as
     * one message because some lpd's are broken and require that
     * they reap the message in one read(2)).  then connect to
     * the lpd, send the request and read the reply.
     */

    req = make_queue_request(client_get_printer(), user_list);

    sock_fd = client_open(rlpr_rlpq->timeout);
    if (sock_fd == -1)
	msg(R_FATAL, 0, "client_open(): cannot connect to lpd");

    if (client_command_noack(sock_fd, req, rlpr_rlpq->timeout, "queue request")
	== 0)
	msg(R_FATAL, 0, "unable to send queue request to lpd");

    if (msg_is_quiet()) {

	char		buffer[BUFSIZ];
	int		pipe_fd[2];
	ssize_t		n_read;

	/*
	 * we'd like to see if there's a "no entries\n" string
	 * returned or not from the lpd.  however, we want to respect
	 * timeouts retrieving this information -- ideally, we'd use
	 * copy_file() since it knows about timeouts.  what to do?
	 * pipes to the rescue.	 we're so friggin' clever sometimes.
	 */

	if (pipe(pipe_fd) == -1)
	    msg(R_FATAL, errno, "pipe");

	if (copy_file(sock_fd, pipe_fd[1], rlpr_rlpq->timeout, 0, "pipe") == 0)
	    msg(R_FATAL, 0, "unable to get queue output from lpd");

	if ((n_read = read(pipe_fd[0], buffer, sizeof buffer - 1)) == -1)
	    msg(R_FATAL, errno, "read of piped data from lpd");

	buffer[n_read] = '\0';
	if (strcasecmp("no entries\n", buffer) == 0)
	    retval = EXIT_FAILURE;

    } else {

	if (copy_file(sock_fd, STDOUT_FILENO, rlpr_rlpq->timeout, 0, "stdout")
	    == 0)
	    msg(R_FATAL, 0, "unable to get queue output from lpd");
    }

    close(sock_fd);

    if ((comp = component_fini(components)) != 0)
	msg(R_FATAL, 0, "component `%s': fini() failed!", comp->name);

    return retval;
}

static const char *
make_queue_request(const char *printer, char * const *user_list)
{
    unsigned int	request_size = 3; /* "\003\n" */
    char	       *request;
    unsigned int	i;

    request_size += strlen(printer);
    for (i = 0; user_list[i] != 0; i++)
	request_size += strlen(user_list[i]) + 1;

    request = xmalloc(request_size);
    sprintf(request, "%c%s", rlpr_rlpq->long_output ? SENDQL : SENDQS, printer);

    for (i = 0; user_list[i] != 0; i++) {
	strcat(request, " ");
	strcat(request, user_list[i]);
    }

    strcat(request, "\n");
    return request;
}

static int
rlpq_init(void)
{
    struct rlpr_rlpq	rlpr_rlpq_tmpl = { 0 };
    struct passwd      *pwd;

    /*
     * we perform a structure copy on a zero-initialized structure
     * instead of using memset() since memset is a raw bit-oriented
     * function, which means that it will generate potentially
     * incorrect results with non-integer types.
     */

    rlpr_rlpq	= xmalloc(sizeof (struct rlpr_rlpq));
    *rlpr_rlpq	= rlpr_rlpq_tmpl;
    pwd = getpwuid(getuid());

    if (pwd == 0 || pwd->pw_name == 0) {
	msg(R_ERROR, errno, "unable to resolve your username");
	return 0;
    }

    rlpr_rlpq->timeout	= R_TIMEOUT_DEFAULT;
    rlpr_rlpq->user	= xstrdup(pwd->pw_name);
    return 1;
}

static int
rlpq_parse_args(int opt)
{
    switch (opt) {

    case 'l':
	rlpr_rlpq->long_output++;
	break;

    case 'm':
	rlpr_rlpq->mine_only++;
	break;

    case -500:
	msg(R_STDOUT, 0, "usage: %s [-Hprinthost] [-Pprinter] [-Xproxy]"
	    " [OPTIONS]\nplease see the manpage for detailed help",
	    program_name);
	exit(EXIT_SUCCESS);
	break;

    case -501:
	rlpr_rlpq->timeout = strtol(optarg, 0, 0);
	break;

    case 'V':
	msg(R_STDOUT, 0, "version "VERSION" from " __DATE__" "__TIME__
	    " -- meem@gnu.org");
	exit(EXIT_SUCCESS);
	break;

    case EOF:
	break;
    }

    return 1;
}

static struct option rlpq_opts[] = {
    { "help",		0, 0, -500  },
    { "long",		0, 0, 'l'   },
    { "mine",		0, 0, 'm'   },
    { "timeout",	1, 0, -501  },
    { "version",	0, 0, 'V'   },
    { 0,		0, 0,  0    }
};

static const char rlpq_opt_list[] = "lmV";

struct component comp_rlpq = {
    "rlpq", rlpq_init, 0,
    { { rlpq_opts, rlpq_opt_list, rlpq_parse_args } }
};
